optimisation/design/Third_Party_API_Design.md

```diff
@@ -0,0 +1,595 @@
+
+```matlab
+optfun_sqw = OptFunction(@user_model_sqw, 'type', 'double', 'inputs', 4, 'parameters', parvec, 'arguments', argcell);
+optfun_dsp = OptFunction(@user_model_dsp, 'type', 'double', 'inputs', 3, 'parameters', parvec, 'arguments', argcell);
+optfun_pow = OptFunction(@user_model_pow, 'type', 'double', 'inputs', 2, 'parameters', parvec, 'arguments', argcell);
+optfun_1d = OptFunction(@user_model_1d, 'type', 'double', 'inputs', 1, 'parameters', parvec, 'arguments', argcell);
+optfun_0d = OptFunction(@user_model_0d, 'type', 'double', 'inputs', 0, 'parameters', parvec, 'arguments', argcell);
+```
+
+where `parvec` is a vector of parameters and `argcell` is a cell-array of arguments (which may be omitted),
+which is passed (as `varargin`) unmodified directly to the user functions.
+
+
+## <a name="python_functions"></a> Python
+
+When a Python function is to be called from Horace running in Matlab, the following signatures should be used:
+
+```python
+def user_model_sqw(qh, qk, ql, en, parameters, *args):
+def user_model_dsp(qh, qk, ql, parameters, *args):
+def user_model_pow(modQ, en, parameters, *args):
+def user_model_1d(en, parameters, *args):
+def user_model_0d(parameters, *args):
+```
+
+A Matlab cell array is passed to Python as a list which corresponds to the positional arguments `*args`.
+The "pythonic" way would be to used named keyword arguments `**kwargs` which are Python `dict`s / Matlab `struct`s,
+however this behaviour is not as natural in Matlab.
+
+In the case of `pyHorace` which is run from Python, slightly modified function signatures are possible:
+
+```python
+def user_model_sqw(qh, qk, ql, en, parameters, *args, **kwargs):
+```
+
+etc. could be used instead, since `pyHorace` will have Python wrappers which can interpret "(string, param)" pairs in a cell array as keyword arguments.
+To set up the `OptFunction` objects in `pyHorace` either keyword or positional arguments will be accepted, but not a mix:
+
+```python
+optfun_sqw = m.OptFunction(user_model_sqw, 'type', 'double', 'inputs', 4, 'parameters', parvec, 'arguments', arglist)
+optfun_sqw = m.OptFunction(user_model_sqw, type='double', inputs=4, parameters=parvec, **kwargs_for_func)
+```
+
+A mix of keywords and Matlab-style pairs of keyword and positional arguments will give an error, as this mix can cause ambiguities which are hard to parse.
+
+
+## <a name="c_functions"></a> C
+
+As we are using `loadlibrary` / `calllib` to get Matlab to evaluate a function in a compiled library,
+we need to use C-calling conventions for these functions for all languages: C, C++ and Fortran.
+In order to accommodate Fortran, these functions *should pass by reference only*, and have no return values.
+
+The function signatures in C is thus:
+
+```c
+void user_model_sqw(const double *qh, const double *qk, const double *ql, const double *en,
+                    const double *parameters, double *results, int *n_elem);
```

Comment:
```suggestion
                    const double *parameters, double *results, unsigned long *n_elem);
```
`n_elem` might be *very* large, it should be at least an unsigned 32-bit integer (2<sup>32</sup>-1 ~ 4.29 billion).

Also, similar to above, we should define signatures for the N<=4 dimensional user functions of type 1.